### ISSUE TITLE
MultipleTimeSeriesCV, why does it iterate backwards in time?

### ISSUE THREAD

**Issue by carstenf** [stefan-jansen/machine-learning-for-trading#314]
It looks like the splits are the wrong way around and should be reversed.
The first split will be used first to compute, than the second split should not use information from the first calculation.
This looks like the other way around, please see the plot.

I made a small pice of code for plotting:

    `
      import matplotlib.pyplot as plt
      import numpy as np
      import pandas as pd
    
    # Initialize the MultipleTimeSeriesCV with your parameters
    cv = MultipleTimeSeriesCV(n_splits=5, train_period_length=300, test_period_length=300, lookahead=50)
    
    # Define the colors for different sections of the plot
    train_color = 'blue'  # In-sample data color
    test_color = 'orange'    # Out-of-sample data color
    lookahead_color = 'green'  # Lookahead period color
    
    # Create subplots
    fig, axs = plt.subplots(cv.n_splits, figsize=(10, 15), sharex=True)
    
    # Iterate over each split
    for i, (train_index, test_index) in enumerate(cv.split(data)):
        train = data.iloc[train_index]
        test = data.iloc[test_index]
    
        # Plot training set
        axs[i].plot(train.loc['AAPL'].index, train.loc['AAPL', 'close'], color=train_color, label='Training Set')
        
        # Plot testing set
        axs[i].plot(test.loc['AAPL'].index, test.loc['AAPL', 'close'], color=test_color, label='Testing Set')
    
        # Highlight the lookahead period
        if len(train) > 0 and len(test) > 0:
            lookahead_start_date = train.loc['AAPL'].index[-1]
            lookahead_end_date = test.loc['AAPL'].index[0]
            axs[i].axvspan(lookahead_start_date, lookahead_end_date, color=lookahead_color, alpha=0.3, label='Lookahead Period')
    
        axs[i].legend(loc='best')
        
        # Formatting the plot
        axs[i].set_title(f'Split {i+1}')
        axs[i].set_xlabel('Date')
        axs[i].set_ylabel('Close Price')
    
    plt.tight_layout()
    plt.show()

`

![image](https://github.com/stefan-jansen/machine-learning-for-trading/assets/58643140/e4ca0653-3de9-471e-82a3-3f3693dbb436)

**Comment by carstenf** [comment]
I probably found a solution, but not fully tested:


    class MultipleTimeSeriesCV:
         """Generates tuples of train_idx, test_idx pairs
         Assumes the MultiIndex contains levels 'symbol' and 'date'
         purges overlapping outcomes"""
    
        def __init__(self,
                     n_splits=3,
                     train_period_length=126,
                     test_period_length=21,
                     lookahead=None,
                     date_idx='date',
                     shuffle=False):
            self.n_splits = n_splits
            self.lookahead = lookahead
            self.test_length = test_period_length
            self.train_length = train_period_length
            self.shuffle = shuffle
            self.date_idx = date_idx
    
    
        def split(self, X, y=None, groups=None):
                unique_dates = X.index.get_level_values(self.date_idx).unique()
                days = sorted(unique_dates)  # Ascending order
                split_idx = []
                for i in range(self.n_splits):
                    # Calculate split indices based on ascending order of days 
                    train_start_idx = i * self.test_length   
                    train_end_idx = train_start_idx + self.train_length
                    test_start_idx = train_end_idx + (self.lookahead or 0)
                    test_end_idx = test_start_idx + self.test_length
            
                    # Ensure we do not exceed the length of days
                    if test_end_idx >= len(days):
                        break
                    
                    split_idx.append((train_start_idx, train_end_idx, test_start_idx, test_end_idx))
            
                dates = X.reset_index()[[self.date_idx]]
            
                for train_start, train_end, test_start, test_end in split_idx:
                    # Adjust the condition to select the right slice based on sorted ascending days
                    train_idx = dates[(dates[self.date_idx] >= days[train_start]) & 
                                      (dates[self.date_idx] < days[train_end])].index
                    test_idx = dates[(dates[self.date_idx] >= days[test_start]) & 
                                     (dates[self.date_idx] < days[test_end])].index
            
                    if self.shuffle:
                        train_idx = np.random.permutation(train_idx)
                    
                    yield train_idx.to_numpy(), test_idx.to_numpy()
    
        def get_n_splits(self, X, y, groups=None):
            return self.n_splits
   
 
the new result:
![image](https://github.com/stefan-jansen/machine-learning-for-trading/assets/58643140/3b8c9fc6-9766-4442-9bb3-d2b4fcb5cffd)

**Comment by elsdes3** [comment]
I think the splits are generated in reverse for two reasons

1. for convenience when implementing the splitter
    - it is straightforward to walk backwards from the start of the unseen data (test split) to capture the training data while also taking into account the lookahead period (or gap)
2. to avoid data leakage (lookahead bias) from the training data into the unseen (test) data

By doing this, it makes it easy to efficiently write the `MultipleTimeSeriesCV` splitter and also for most users to quickly read and understand the logic.

Also, even though the splits are generated in reverse order, this splitter can be used in time-series cross-validation without any problems.

When trying the visualize (plot) the outputs using the splitter and iterate forwards in time, just iterate over the splits in reverse and the plots will be shown in chronological order
```python
# Iterate over each split
for i, (train_index, test_index) in enumerate(cv.split(data)[::-1]):
```

**Comment by elsdes3** [comment]
> It looks like the splits are the wrong way around and should be reversed.

I think the splits are generated in reverse for convenience when implementing the splitter. It is straightforward to walk backwards from the start of the unseen data (test split) to capture the training data while also taking into account the lookahead period (or gap) and avoiding data leakage (lookahead bias). So, by generating the splits in reverse order, it makes it easy for most users to quickly read and understand the logic.

> The first split will be used first to compute, than the second split should not use information from the first calculation. This looks like the other way around,

Can you clarify what you mean by this? For each split, the training data is always separated in time from the test data.

> I made a small pice of code for plotting:

You can just iterate over the splits in reverse and the plots will be shown in chronological order
```python
# Iterate over each split
for i, (train_index, test_index) in enumerate(cv.split(data)[::-1]):
```

Note: Even though the splits are generated in reverse order, this splitter can be used in time-series cross-validation without any data leakage.

**Comment by carstenf** [comment]
First, many thanks for reviewing. 
I understood that data I used before for training/testing is „polluted“ with information and should not be used for further tests. For training should be ok.
In the original case the train data from the first run is used as test data in the second run. 
As I understood this should be a no go, maybe I’m wrong. 
If Splitt 3 was the first run, Splitt 2 the second run and so on, it should be ok…

**Comment by stefan-jansen** [comment]
The folds are just generated in reverse order because the assumption is you want to validate on the most recent N periods. The data that go into training/validating one model do not create lookahead bias for another model trained on different folds.

**Comment by elsdes3** [comment]
> ...In the original case the train data from the first run is used as test data in the second run. As I understood this should be a no go, maybe I’m wrong...

That is a good point you make, but it is only partially true. Some of the training data from the first split is indeed used as the test data in the second split. However, all the data is not used as test data in the second split and that is the most important part. This is how sliding window timeseries splitters are defined (see the left hand side of figure 4 from [here](https://eng.uber.com/forecasting-introduction/)). Even when that happens, it is still not a problem because the test data for the second split is preceeded by the training data so there is no lookahead bias. This is actually evident in Split 2 in [your first plot](https://github.com/stefan-jansen/machine-learning-for-trading/issues/314#issue-2086023661).

So, it should be fine to use this splitter without lookahead bias for timeseries cross-validation.
